File: examples/apps/riffer/go/llm/transform.go
//go:build nxuskit

// Package llm provides LLM-powered features for riffer.
package llm

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/nxus-SYSTEMS/nxusKit/examples/apps/riffer"
	nxuskit "github.com/nxus-SYSTEMS/nxusKit/packages/nxuskit-go"
)

// TransformRequest represents the parsed transformation parameters
type TransformRequest struct {
	Transpose   *TransposeParams  `json:"transpose,omitempty"`
	Tempo       *TempoParams      `json:"tempo,omitempty"`
	Invert      *InvertParams     `json:"invert,omitempty"`
	Retrograde  *RetrogradeParams `json:"retrograde,omitempty"`
	Augment     *AugmentParams    `json:"augment,omitempty"`
	Diminish    *DiminishParams   `json:"diminish,omitempty"`
	KeyChange   *KeyChangeParams  `json:"key_change,omitempty"`
	Description string            `json:"description,omitempty"`
}

// TransposeParams for transpose transformation
type TransposeParams struct {
	Semitones int `json:"semitones"`
}

// TempoParams for tempo change
type TempoParams struct {
	BPM int `json:"bpm"`
}

// InvertParams for inversion
type InvertParams struct {
	Pivot *int `json:"pivot,omitempty"`
}

// RetrogradeParams for retrograde
type RetrogradeParams struct {
	Enabled bool `json:"enabled"`
}

// AugmentParams for augmentation
type AugmentParams struct {
	Factor float64 `json:"factor"`
}

// DiminishParams for diminution
type DiminishParams struct {
	Factor float64 `json:"factor"`
}

// KeyChangeParams for key change
type KeyChangeParams struct {
	Root string `json:"root"`
	Mode string `json:"mode"`
}

const transformSystemPrompt = `You are a music transformation assistant. Given a natural language description of a musical transformation, you will respond with a JSON object containing the transformation parameters.

Available transformations:
- "transpose": Shift all notes by semitones. Use "semitones" field (-12 to +12).
  Examples: "up a major third" = +4, "down a perfect fifth" = -7, "up an octave" = +12
- "tempo": Change the tempo. Use "bpm" field (20-300).
- "invert": Mirror the melody around a pivot pitch. Use "pivot" field (MIDI note, 0-127, or null for first note).
- "retrograde": Reverse the note order. Use "enabled" field (true/false).
- "augment": Stretch durations. Use "factor" field (e.g., 2.0 = twice as long).
- "diminish": Compress durations. Use "factor" field (e.g., 2.0 = half as long).
- "key_change": Change key. Use "root" (C, C#, D, etc.) and "mode" (major/minor) fields.

Respond ONLY with a JSON object. Example:
{"transpose": {"semitones": 7}, "description": "Transposed up a perfect fifth"}

For multiple transformations:
{"transpose": {"semitones": 5}, "tempo": {"bpm": 90}, "description": "Transposed up and slowed down"}

If the prompt is unclear, respond with your best interpretation.`

// TransformWithPrompt transforms a sequence based on a natural language prompt.
// It uses an LLM to interpret the prompt and apply the appropriate transformations.
func TransformWithPrompt(ctx context.Context, seq *riffer.Sequence, prompt string) (string, error) {
	keyStr := "Unknown"
	if seq.Context.KeySignature != nil {
		keyStr = seq.Context.KeySignature.String()
	}

	userPrompt := fmt.Sprintf(
		"Transform request: \"%s\"\n\nCurrent sequence info:\n- Notes: %d\n- Key: %s\n- Tempo: %d BPM",
		prompt,
		len(seq.Notes),
		keyStr,
		seq.Context.Tempo,
	)

	// Call LLM
	response, err := callLLM(ctx, transformSystemPrompt, userPrompt)
	if err != nil {
		return "", err
	}

	// Extract JSON from response (might be in code blocks)
	jsonStr := extractJSON(response)

	// Parse the response
	var req TransformRequest
	if err := json.Unmarshal([]byte(jsonStr), &req); err != nil {
		return "", fmt.Errorf("failed to parse LLM response: %w. Response: %s", err, response)
	}

	// Apply transformations
	var applied []string

	if req.Transpose != nil {
		if err := riffer.Transpose(seq, int8(req.Transpose.Semitones)); err != nil {
			return "", fmt.Errorf("transpose failed: %w", err)
		}
		applied = append(applied, fmt.Sprintf("Transposed by %d semitones", req.Transpose.Semitones))
	}

	if req.Tempo != nil {
		if err := riffer.ChangeTempo(seq, uint16(req.Tempo.BPM)); err != nil {
			return "", fmt.Errorf("tempo change failed: %w", err)
		}
		applied = append(applied, fmt.Sprintf("Changed tempo to %d BPM", req.Tempo.BPM))
	}

	if req.Invert != nil {
		var pivot *uint8
		if req.Invert.Pivot != nil {
			p := uint8(*req.Invert.Pivot)
			pivot = &p
		}
		if err := riffer.Invert(seq, pivot); err != nil {
			return "", fmt.Errorf("invert failed: %w", err)
		}
		applied = append(applied, "Inverted melody")
	}

	if req.Retrograde != nil && req.Retrograde.Enabled {
		if err := riffer.Retrograde(seq); err != nil {
			return "", fmt.Errorf("retrograde failed: %w", err)
		}
		applied = append(applied, "Applied retrograde")
	}

	if req.Augment != nil {
		if err := riffer.Augment(seq, float32(req.Augment.Factor)); err != nil {
			return "", fmt.Errorf("augment failed: %w", err)
		}
		applied = append(applied, fmt.Sprintf("Augmented by factor %v", req.Augment.Factor))
	}

	if req.Diminish != nil {
		if err := riffer.Diminish(seq, float32(req.Diminish.Factor)); err != nil {
			return "", fmt.Errorf("diminish failed: %w", err)
		}
		applied = append(applied, fmt.Sprintf("Diminished by factor %v", req.Diminish.Factor))
	}

	if req.KeyChange != nil {
		targetKey, ok := riffer.ParseKey(req.KeyChange.Root + modeToSuffix(req.KeyChange.Mode))
		if !ok {
			return "", fmt.Errorf("invalid key: %s %s", req.KeyChange.Root, req.KeyChange.Mode)
		}
		if err := riffer.KeyChange(seq, *targetKey); err != nil {
			return "", fmt.Errorf("key change failed: %w", err)
		}
		applied = append(applied, fmt.Sprintf("Changed key to %s", targetKey.String()))
	}

	// Build description
	if len(applied) == 0 {
		if req.Description != "" {
			return req.Description, nil
		}
		return "No transformations applied", nil
	}

	return strings.Join(applied, "; "), nil
}

func modeToSuffix(mode string) string {
	mode = strings.ToLower(mode)
	if strings.Contains(mode, "minor") {
		return "m"
	}
	return ""
}

func extractJSON(response string) string {
	// Try to find JSON in code blocks
	if idx := strings.Index(response, "```json"); idx != -1 {
		start := idx + 7
		if end := strings.Index(response[start:], "```"); end != -1 {
			return strings.TrimSpace(response[start : start+end])
		}
	}
	if idx := strings.Index(response, "```"); idx != -1 {
		start := idx + 3
		if end := strings.Index(response[start:], "```"); end != -1 {
			content := strings.TrimSpace(response[start : start+end])
			// Skip language identifier if present
			if newline := strings.Index(content, "\n"); newline != -1 {
				return strings.TrimSpace(content[newline+1:])
			}
		}
	}
	return strings.TrimSpace(response)
}

func callLLM(ctx context.Context, systemPrompt, userPrompt string) (string, error) {
	// Try providers in order: Claude, OpenAI, Ollama
	if apiKey := os.Getenv("ANTHROPIC_API_KEY"); apiKey != "" {
		return callClaude(ctx, apiKey, systemPrompt, userPrompt)
	}

	if apiKey := os.Getenv("OPENAI_API_KEY"); apiKey != "" {
		return callOpenAI(ctx, apiKey, systemPrompt, userPrompt)
	}

	// Try Ollama as fallback
	return callOllama(ctx, systemPrompt, userPrompt)
}

func callClaude(ctx context.Context, apiKey, systemPrompt, userPrompt string) (string, error) {
	provider, err := nxuskit.NewClaudeFFIProvider(
		nxuskit.WithClaudeAPIKey(apiKey),
	)
	if err != nil {
		return "", fmt.Errorf("failed to create Claude provider: %w", err)
	}

	req := &nxuskit.ChatRequest{
		Model: "claude-haiku-4-5-20251001",
		Messages: []nxuskit.Message{
			nxuskit.SystemMessage(systemPrompt),
			nxuskit.UserMessage(userPrompt),
		},
	}

	resp, err := provider.Chat(ctx, req)
	if err != nil {
		return "", fmt.Errorf("Claude API error: %w", err)
	}

	return resp.Content, nil
}

func callOpenAI(ctx context.Context, apiKey, systemPrompt, userPrompt string) (string, error) {
	provider, err := nxuskit.NewOpenAIFFIProvider(
		nxuskit.WithOpenAIAPIKey(apiKey),
	)
	if err != nil {
		return "", fmt.Errorf("failed to create OpenAI provider: %w", err)
	}

	req := &nxuskit.ChatRequest{
		Model: "gpt-4o-mini",
		Messages: []nxuskit.Message{
			nxuskit.SystemMessage(systemPrompt),
			nxuskit.UserMessage(userPrompt),
		},
	}

	resp, err := provider.Chat(ctx, req)
	if err != nil {
		return "", fmt.Errorf("OpenAI API error: %w", err)
	}

	return resp.Content, nil
}

func callOllama(ctx context.Context, systemPrompt, userPrompt string) (string, error) {
	provider, err := nxuskit.NewOllamaFFIProvider()
	if err != nil {
		return "", fmt.Errorf("failed to create Ollama provider: %w", err)
	}

	req := &nxuskit.ChatRequest{
		Model: "llama3.2",
		Messages: []nxuskit.Message{
			nxuskit.SystemMessage(systemPrompt),
			nxuskit.UserMessage(userPrompt),
		},
	}

	resp, err := provider.Chat(ctx, req)
	if err != nil {
		return "", fmt.Errorf("no LLM available. Set ANTHROPIC_API_KEY or OPENAI_API_KEY, or run Ollama locally. Error: %w", err)
	}

	return resp.Content, nil
}

File: examples/apps/riffer/rust/llm/transform.rs
//! Natural Language Music Transformation
//!
//! Uses LLMs to interpret natural language prompts and generate
//! transformation instructions for music sequences.

use std::env;

use nxuskit::builders::{ClaudeProvider, OllamaProvider, OpenAIProvider};
use nxuskit::{ChatRequest, Message};

use crate::engine::{augment, change_tempo, diminish, invert, key_change, retrograde, transpose};
use crate::errors::{Result, RifferError};
use crate::types::{KeySignature, Mode, PitchClass, Sequence};

/// Transform a sequence based on a natural language prompt
///
/// # Arguments
/// * `sequence` - The music sequence to transform (modified in place)
/// * `prompt` - Natural language description of the desired transformation
///
/// # Example prompts:
/// - "Transpose up a perfect fifth"
/// - "Make it twice as slow"
/// - "Convert to A minor"
/// - "Invert the melody"
/// - "Play it backwards"
pub async fn transform_with_prompt(sequence: &mut Sequence, prompt: &str) -> Result<String> {
    // Build the system prompt for transformation interpretation
    let system_prompt = r#"You are a music transformation assistant. Given a natural language description of a musical transformation, you will respond with a JSON object containing the transformation parameters.

Available transformations:
- "transpose": Shift all notes by semitones. Use "semitones" field (-12 to +12).
  Examples: "up a major third" = +4, "down a perfect fifth" = -7, "up an octave" = +12
- "tempo": Change the tempo. Use "bpm" field (20-300).
- "invert": Mirror the melody around a pivot pitch. Use "pivot" field (MIDI note, 0-127, or null for first note).
- "retrograde": Reverse the note order. Use "enabled" field (true/false).
- "augment": Stretch durations. Use "factor" field (e.g., 2.0 = twice as long).
- "diminish": Compress durations. Use "factor" field (e.g., 2.0 = half as long).
- "key_change": Change key. Use "root" (C, C#, D, etc.) and "mode" (major/minor) fields.

Respond ONLY with a JSON object. Example:
{"transpose": {"semitones": 7}, "description": "Transposed up a perfect fifth"}

For multiple transformations:
{"transpose": {"semitones": 5}, "tempo": {"bpm": 90}, "description": "Transposed up and slowed down"}

If the prompt is unclear, respond with your best interpretation."#;

    let user_prompt = format!(
        "Transform request: \"{}\"\n\nCurrent sequence info:\n- Notes: {}\n- Key: {}\n- Tempo: {} BPM",
        prompt,
        sequence.notes.len(),
        sequence
            .context
            .key_signature
            .as_ref()
            .map(|k| k.to_string())
            .unwrap_or_else(|| "Unknown".to_string()),
        sequence.context.tempo
    );

    // Get LLM response
    let response = call_llm(system_prompt, &user_prompt).await?;

    // Parse the JSON response - try to extract JSON from markdown code blocks
    let json_str = extract_json(&response);
    let parsed: serde_json::Value = serde_json::from_str(json_str).map_err(|e| {
        RifferError::LlmParseError(format!(
            "Failed to parse LLM response: {}. Response: {}",
            e, response
        ))
    })?;

    // Apply transformations
    let mut applied = Vec::new();

    if let Some(transpose_obj) = parsed.get("transpose") {
        if let Some(semitones) = transpose_obj.get("semitones").and_then(|s| s.as_i64()) {
            transpose(sequence, semitones as i8)?;
            applied.push(format!("Transposed by {} semitones", semitones));
        }
    }

    if let Some(tempo_obj) = parsed.get("tempo") {
        if let Some(bpm) = tempo_obj.get("bpm").and_then(|b| b.as_u64()) {
            change_tempo(sequence, bpm as u16)?;
            applied.push(format!("Changed tempo to {} BPM", bpm));
        }
    }

    if let Some(invert_obj) = parsed.get("invert") {
        let pivot = invert_obj
            .get("pivot")
            .and_then(|p| p.as_u64())
            .map(|p| p as u8);
        invert(sequence, pivot)?;
        applied.push("Inverted melody".to_string());
    }

    if let Some(retrograde_obj) = parsed.get("retrograde") {
        if retrograde_obj
            .get("enabled")
            .and_then(|e| e.as_bool())
            .unwrap_or(false)
        {
            retrograde(sequence)?;
            applied.push("Applied retrograde".to_string());
        }
    }

    if let Some(augment_obj) = parsed.get("augment") {
        if let Some(factor) = augment_obj.get("factor").and_then(|f| f.as_f64()) {
            augment(sequence, factor as f32)?;
            applied.push(format!("Augmented by factor {}", factor));
        }
    }

    if let Some(diminish_obj) = parsed.get("diminish") {
        if let Some(factor) = diminish_obj.get("factor").and_then(|f| f.as_f64()) {
            diminish(sequence, factor as f32)?;
            applied.push(format!("Diminished by factor {}", factor));
        }
    }

    if let Some(key_obj) = parsed.get("key_change") {
        let root_str = key_obj.get("root").and_then(|r| r.as_str()).unwrap_or("C");
        let mode_str = key_obj
            .get("mode")
            .and_then(|m| m.as_str())
            .unwrap_or("major");

        let root = parse_pitch_class(root_str).unwrap_or(PitchClass::C);
        let mode = if mode_str.to_lowercase().contains("minor") {
            Mode::Minor
        } else {
            Mode::Major
        };

        let target_key = KeySignature::new(root, mode);
        key_change(sequence, &target_key)?;
        applied.push(format!("Changed key to {}", target_key));
    }

    // Build result description
    let description = if applied.is_empty() {
        parsed
            .get("description")
            .and_then(|d| d.as_str())
            .unwrap_or("No transformations applied")
            .to_string()
    } else {
        applied.join("; ")
    };

    Ok(description)
}

/// Extract JSON from a response that might be wrapped in markdown code blocks
fn extract_json(response: &str) -> &str {
    // Try to find JSON in code blocks
    if let Some(start) = response.find("```json") {
        let start = start + 7;
        if let Some(end) = response[start..].find("```") {
            return response[start..start + end].trim();
        }
    }
    if let Some(start) = response.find("```") {
        let start = start + 3;
        if let Some(end) = response[start..].find("```") {
            let content = response[start..start + end].trim();
            // Skip the language identifier if present
            if let Some(newline) = content.find('\n') {
                return content[newline + 1..].trim();
            }
        }
    }
    // Return as-is if no code blocks found
    response.trim()
}

/// Parse a pitch class string like "C", "C#", "Db", etc.
fn parse_pitch_class(s: &str) -> Option<PitchClass> {
    match s.to_uppercase().as_str() {
        "C" => Some(PitchClass::C),
        "C#" | "DB" => Some(PitchClass::Cs),
        "D" => Some(PitchClass::D),
        "D#" | "EB" => Some(PitchClass::Ds),
        "E" => Some(PitchClass::E),
        "F" => Some(PitchClass::F),
        "F#" | "GB" => Some(PitchClass::Fs),
        "G" => Some(PitchClass::G),
        "G#" | "AB" => Some(PitchClass::Gs),
        "A" => Some(PitchClass::A),
        "A#" | "BB" => Some(PitchClass::As),
        "B" => Some(PitchClass::B),
        _ => None,
    }
}

/// Call an LLM provider to get a response
async fn call_llm(system_prompt: &str, user_prompt: &str) -> Result<String> {
    // Try providers in order: Claude, OpenAI, Ollama
    if let Ok(api_key) = env::var("ANTHROPIC_API_KEY") {
        return call_claude(&api_key, system_prompt, user_prompt).await;
    }

    if let Ok(api_key) = env::var("OPENAI_API_KEY") {
        return call_openai(&api_key, system_prompt, user_prompt).await;
    }

    // Try Ollama as fallback (no API key needed)
    call_ollama(system_prompt, user_prompt).await
}

async fn call_claude(api_key: &str, system_prompt: &str, user_prompt: &str) -> Result<String> {
    let provider = ClaudeProvider::builder()
        .api_key(api_key)
        .model("claude-haiku-4-5-20251001")
        .build()
        .map_err(|e| RifferError::LlmUnavailable(e.to_string()))?;

    let request = ChatRequest::new("claude-haiku-4-5-20251001")
        .with_message(Message::system(system_prompt))
        .with_message(Message::user(user_prompt));

    let response = provider
        .chat_async(request)
        .await
        .map_err(|e| RifferError::LlmUnavailable(e.to_string()))?;

    Ok(response.content)
}

async fn call_openai(api_key: &str, system_prompt: &str, user_prompt: &str) -> Result<String> {
    let provider = OpenAIProvider::builder()
        .api_key(api_key)
        .model("gpt-4o-mini")
        .build()
        .map_err(|e| RifferError::LlmUnavailable(e.to_string()))?;

    let request = ChatRequest::new("gpt-4o-mini")
        .with_message(Message::system(system_prompt))
        .with_message(Message::user(user_prompt));

    let response = provider
        .chat_async(request)
        .await
        .map_err(|e| RifferError::LlmUnavailable(e.to_string()))?;

    Ok(response.content)
}

async fn call_ollama(system_prompt: &str, user_prompt: &str) -> Result<String> {
    let provider = OllamaProvider::builder()
        .model("llama3.2")
        .build()
        .map_err(|e| RifferError::LlmUnavailable(e.to_string()))?;

    let request = ChatRequest::new("llama3.2")
        .with_message(Message::system(system_prompt))
        .with_message(Message::user(user_prompt));

    let response = provider.chat_async(request).await.map_err(|e| {
        RifferError::LlmUnavailable(format!(
            "No LLM available. Set ANTHROPIC_API_KEY or OPENAI_API_KEY, or run Ollama locally. Error: {}",
            e
        ))
    })?;

    Ok(response.content)
}
